[chore] AWS S3 기본 설정

epicblues opened the pull request:
기본적으로 한빈님이 작성한 File 관련 클래스들을 그대로 사용했습니다.
기존과 다른 사항들만 아래에서 확인해주시면 될 것 같습니다!

## 특이 사항

### AWS  S3 테스트 관련
* 매번 단위, 통합 테스트마다 s3 관련 의존성을 로드하고, 실제로 s3에 이미지를 전송하는 것이 비효율적이라고 판단했습니다.
* 따라서 **테스트 환경에서는 s3 관련 의존성을 로드하지 않도록 설정했습니다**. `build.gradle` 참고해주세요
* 기존에 s3 라이브러리에 multipart 전송을 책임지던 `S3FileStore`를 테스트 환경에서는 빈으로 등록하지 않습니다.
* 대신 같은 인터페이스를 가진 테스트 전용 클래스인 `FileStoreMock`를 빈으로 등록합니다.
* 이미지 메시지 전송 여부를 테스트 하고 싶으시면, 기존처럼 `@MockBean`을 통해 FileStore Mock을 받아 Mockito 방법을 활용하시면 될 것 같습니다.

### AWS 관련 비밀 설정

* `application-aws.yml`을 `.gitignore`에 추가했습니다. (위치 src/main/java/resources )
* 대충 이런 형식입니다.
```yaml
cloud:
  aws:
    credentials:
      access-key: 
      secret-key: 

    region:
      static: 
    s3:
      bucket: 
    stack:
      auto: false
logging: ## SPRING CLOUD AWS 에서 EC2 관련 메타데이터 로깅 기능 비활성화
  level:
    com:
      amazonaws:
        util:
          EC2MetadataUtils: error
```
* 위치를 보면 아시겠지만 테스트 환경과는 무관합니다. 
* 로컬에서도 postman으로 이미지 업로드 e2e 테스트를 실행하는 상황이 나올 것 같아서 다른 secret처럼 환경 변수 처리하지 않았습니다. 
* **`.gitignore`된 파일은 슬랙이나 다른 메시지를 통해 공유하겠습니다. 필요하면 말씀해주세요**

### S3Config 제거
* S3Config 없이도 파일 전송이 정상적으로 되는 것을 확인했습니다.
* 혹시 예상되는 문제 있으면 다시 복구시키겠습니다.

epicblues committed:
[chore] AWS S3 기본 설정

epicblues committed:
[refactor] SpyFileStore spy 메서드

epicblues committed:
[refactor] S3Config 제거

s3 config 없이도 파일 전송에 문제 없음

res-cogitans commented on src/test/java/com/prgrms/rg/infrastructure/file/SpyFileStore.java:
```diff
@@ -0,0 +1,62 @@
+	/*
+	 * 이미지 전송 요청이 제대로 전달 되었는가 확인하는 메서드
+	 */
+	public void assertSaveCommandCalledOnce(String url) {
```
이건 Mockito 사용하지 않고, Mockito.verify 로직을 구현한 거죠?

res-cogitans replied:
assertion이 테스트 메서드 안에 있지 않으니까 무언가 어색한? 느낌은 드네요. (이건 제가 알못이라 그런 부분이라 생각합니다) 요건 리뷰라기 보다는 개인적인 궁금증인데, 민성님이 코드를 작성하신 의도가 궁금합니다. Mockito 의존성을 사용하지 않기 위함인가요?

epicblues replied:
넵 단위 테스트에서 배운 내용 (321page 참고)을 활용해봤습니다. 책에서는 플루언트 인터페이스를 사용하는데, FileStore 검증에서는 필요성을 느끼지 못해서 적용하지 않았습니다.

다른 이유를 살펴보면, 목 프레임워크를 사용할 때 stub 관련 코드를 작성하는 것이 테스트 코드의 가독성을 해치고, 유지보수하기 힘든 테스트코드로 만든다고 느꼈습니다. FileStore의 경우에는 url을 반환하는 save 메서드를 예시로 들 수 있습니다.

`given(mockFileStore.save(multipartFile, fileName)).willReturn( PREFIX_URL + fileName); `
`verify(mockFileStore, times(1)).save(multipartFile, fileName);`

하지만 결국 SpyFileStore도 FileStore 인터페이스가 수정되면 같이 수정되어야 할 코드이므로 유지 보수 문제는 크게 해결되지 않는 것 같네요. 가독성 문제도 제 기준에서 더 나아 보일 뿐이지 한빈님처럼 오히려 어색하게 느껴져서 사용에 혼란을 초래할 수도 있구요. 다른 팀원들도 한빈님과 비슷한 생각일 경우 검증 메서드는 지우는 것이 맞겠네요.

blessing333 replied:
저도 처음보는 구조긴 하네요. 
생소하긴 해도 문제될 부분은 일단 없다고 생각해요
다만 saveCommandCache, deleteCommandCache는 접근 제어자를 private로 설정하는건 어떨까요??

컬렉션들의 접근이 열려있어서 민성님이 작성하신 assert 관련 메소드들을 사용 안하고 컬렉션을 이용해서 직접 검증을 해버리는 경우도 있을것같아요!

epicblues replied:
앗 그러네요! 이런 기본적인 실수를... 바로 고치겠습니다

blessing333 replied:
추가로 궁금한점이 있는데, 각 테스트마다 saveCommandCache, deleteCommandCache는 초기화 되나요?? 
빈으로 등록된 순간 싱글톤으로 관리되어 전역에서 공통으로 사용될 것 같은데 각 테스트 실행마다 빈들이 초기화 됐었는지 햇갈리네요

epicblues replied:
좀있다 확인해보겠습니다 개인적인 추측으로는 @Order 를 적용하는 테스트가 아닌 이상 테스트메서드마다 초기화될것같은데 그렇지 않을 경우 사용하지 않는 게 맞겠네요

epicblues replied:
@blessing333 님이 우려하신 대로 전역으로 사용되고 있었네요.🙄 
@Scope("prototype")를 적용해보니까 해결되긴 했는데...  왜 Mockito 같은 테스트 프레임워크를 사용하는지 알겠습니다.

epicblues replied:
좀 더 살펴보니 FileStoreSpy의 scope를 prototype을 바뀌어도 이것을 의존하는 MultipartFileManager가 singleton인 이상 의미가 없을 것 같아 FileStoreSpy를 Dummy 클래스로 바꿨습니다. 실제 상호작용을 검증하고 싶을 때는 @MockBean을 통해 FIleStore 인터페이스 Mock을 DI 받아 사용해야 할 것 같습니다. 
나중에 어썸한 해결책이 나오면 다시 도전해보겠습니다.

res-cogitans commented on src/test/java/com/prgrms/rg/infrastructure/file/SpyFileStore.java:
```diff
@@ -0,0 +1,62 @@
+		assertThat(multipartFile).isNotNull();
+		var url = prefix + fileName;
+		if(saveCommandCache.containsKey(url)) {
+			saveCommandCache.put(url, saveCommandCache.get(url) + 1);
```
정말 사소한 부분이긴 합니다만 이 부분은 getOrDefault() 이용하여 다음과 같이 if문 사용하지 않고 축약 가능합니다. 요즘 코테 공부할 때 안 건데 유용하더라구요.
```java
saveCommandCache.put(url, saveCommandCache.getOrDefault(url, 0) +1);

epicblues replied:
오 이렇게 쓰는 거였군요. 맨날 쓰려고 시도해보다가 실패했었는데....🙄 적용해보겠습니다!

res-cogitans commented on src/test/java/com/prgrms/rg/infrastructure/file/SpyFileStore.java:
```diff
@@ -0,0 +1,62 @@
+
+	@Override
+	public String save(MultipartFile multipartFile, String fileName) throws FileIOException {
+		assertThat(multipartFile).isNotNull();
```
이 부분은 아직 저희가 테스트를 작성하진 않은 상황이라 괜한 걱정일지도 모르겠지만,

정말 만약에 테스트 중, 요청 `multipartFile`이 `null`인 방식으로 테스트를 수행하는 경우가 있다면,
(예를 들어, 라이딩 게시글을 작성하는 데, 어떤 경우에는 사진 데이터를 넣지 않고 그냥 글을 올릴 수도 있으니까요)

그 테스트의 경우, 실제 운영상황이라면 운영 코드에서는 파일 값이 `null`이라고 해서 Exception을 던지지 않기 때문에 통과해야 정상인 상황임에도,
테스트용 SpyFileStore의 null처리 구현방식(assert문)으로 인해, 테스트가 부분 실패로 나오게 될 것 같은데,
(위의 SpyFileStore에서 assertion이 실패하니까요)

이 부분은 변경이 필요하지 않을까 싶습니다.

epicblues replied:
엇 저는 한빈님이 구현하신 `S3FileStore.save`에서 `multipartFile`이 null일 경우 null pointer exception이 반드시 발생하기 떄문에 운영 코드와 테스트 코드가 모순되지는 않는다고 보는데, 제가 잘못 생각하는 걸까요?

```java
    @Override
    public String save(MultipartFile multipartFile, String fileName) {
	ObjectMetadata metadata = new ObjectMetadata();
        // multipartFile이 null일 경우 반드시 발생
	metadata.setContentLength(multipartFile.getSize());
	metadata.setContentType(multipartFile.getContentType());
```

epicblues replied:
테스트

blessing333 replied:
아마 multipartFile이 null일 경우에는 FileStore 관련 로직으로 실행 흐름이 안넘어오도록 구현되지 않을까요?

kkj0419 replied:
저번 프로젝트 때 민재님이 이야기한 방식으로 구현되었던 걸로 기억해요!! 서비스 쪽에서 null 체크를 해주었던걸로 기억하고 있어서 이 부분은 저도 크게 문제될 건 없을것 같습니다

res-cogitans replied:
헉 제가 답장이 늦었군요. 죄송합니다.
민재님 말씀대로 FileManager 단에서 null 체크 하고, 실행 로직을 넘기지 않아서 문제될 게 없겠네요.
순간적으로 Manager랑 Store가 헷갈렸네요. FileManager에 테스트 대역을 사용하지 않는다면 문제 없을 것 같아요.
문제 없습니다!

res-cogitans reviewed (commented):
민성님 코드 잘 봤습니다!

특히 테스트용 `SpyFileStore`를 재밌게 봤습니다. 개인적인 의문사항 몇 가지 덧붙여 두었습니다.
- 추가적으로 프로젝트를 진행하면서 예전 `FileManager`에 로직 변경이나 리팩토링이 필요하면 그 때 저나 다른 분들이 추가적으로 수정해줘도 좋을 듯 합니다.
- 리뷰 사항 중에 `SpyFileStore`에서 `multipartFile`이 `null`일 경우 `assertion`이 실패하는 방식만, 다른 방식으로 변경한다면, 저는 머지 하셔도 좋다고 생각합니다.

epicblues committed:
[refactor] SpyFileStore 검증 명령 캐싱 식별자 변경

검증 명령 식별자 (파일 이름 + multipartFile 문자열 직렬화 값)

epicblues committed:
[fix] SpyFileStore 버그 수정

epicblues committed:
[refactor] 이름 변경 SpyFileStore -> FileStoreSpy

epicblues committed:
[chore] FileStoreSpy @Primary 적용

테스트환경에서는 실제 FileStore 구현체를 빈으로 등록시키지 않기 위함

HunkiKim reviewed (approved):
잘 봤습니다! 혹시 카카오 프로필 이미지도 url로 날라오는데 이건 이미지 레포에 저장할까요 아니면 User에 저장할까요?

epicblues commented:
> 잘 봤습니다! 혹시 카카오 프로필 이미지도 url로 날라오는데 이건 이미지 레포에 저장할까요 아니면 User에 저장할까요?

@HunkiKim 이건 아마 User 도메인을 담당하고 있는 @res-cogitans 과 이야기해봐야 할 것 같습니다! 확장성을 고려하지 않았을 때는 (1차 버전에서는 카카오만 사용할 것이므로) 토큰에 이미지 url만 들고 다니게 하는 방법도 있겠네요.

epicblues committed:
[refactor] FileStoreSpy 캡슐화

epicblues committed:
[refactor] FileStoreSpy Scope prototype 적용

테스트 메서드 마다 독립적인 테스트 환경을 제공하기 위해 사용.
일반적인 FileStore 구현체들은 가변 상태가 존재하지 않으나 FileStoreSpy는 호출된 명령을 캐싱하기 때문에 테스트가 끝날 때마다 초기화가 필요

epicblues committed:
[fix] FileStoreSpy -> Dummy

FileStoreSpy의 캐싱된 호출 데이터가 테스트 간에 공유됨에 따라 검증 사용에 잠재적인 위험성이 있어서 Dummy 형태로 대체

res-cogitans commented:
> 잘 봤습니다! 혹시 카카오 프로필 이미지도 url로 날라오는데 이건 이미지 레포에 저장할까요 아니면 User에 저장할까요?

이미지 레포에 저장하는 것을 생각하기는 했습니다. 이 부분은 민성님 말씀대로 일단은 토큰에 들고 다니게 하다가 나중에 바꿔도 괜찮구요.

epicblues committed:
Merge branch 'develop' into 13-chore-s3-관련-설정
